### ISSUE TITLE
NixOS test driver seems to always wait until all test VMs have exited on parallel builds

### ISSUE THREAD

**Issue by Zocker1999NET** [NixOS/nixpkgs#351200]
### Describe the bug
The nixos-test driver seems to have a bug where, in case of parallel test derivation "builds", all NixOS tests finish at the same time, even through a test should have finished way quicker than the other.

### Steps To Reproduce
1. Setup nixos tests with a notable expected difference in runtime (see flake example below)
2. Run both tests in parallel (for better visualization, I chose: `nix flake check --log-format internal-json -v |& nom --json`)
3. Check runtimes of tests

### Expected behavior
The short test should be finished way earlier than the long test.

### Screenshots
![image](https://github.com/user-attachments/assets/818f5074-aae1-48ce-85b3-a34fb0679616)

Both tests required 53 to 54 seconds, even through the short test only required, as visible in the log.

### Additional context
This is the flake I built as an example:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
  };
  outputs = { nixpkgs, self, ... }:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs { inherit system; };
    in
    {
      checks.${system} = {
        short = pkgs.nixosTest {
          name = "short";
          nodes.tested = {};
          testScript = ''
            start_all()
            tested.wait_for_unit("default.target");
            tested.succeed("sleep 2");
          '';
        };
        long = pkgs.nixosTest {
          name = "long";
          nodes.tested = {};
          testScript = ''
            start_all()
            tested.wait_for_unit("default.target");
            tested.succeed("sleep 40");
          '';
        };
      };
    };
}
```

I decided to report this as this:
- was not reported yet (at least I have not found another issue on GitHub already)
- sometimes prevents tests from being "marked as successful" (i.e. nix caching the result of the derivation) if just one of many tests fails, even through the others were already finished, triggering an inefficient re-run of most tests later on

### Notify maintainers

@roberth @K900 
(as this is part of `nixos/lib`, without meta.maintainers, I chose the two top commiters per `git shortlog -s -n nixos/lib/testing`) 

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.57, NixOS, 24.05 (Uakari), 24.05.20241019.a9b86fc`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.8`
 - nixpkgs: `/nix/store/lbqj1cndic4121whnx8xm0jgb1c8x4xx-source`
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc
